File: tools/bin/generate_icon.py
#!/usr/bin/env python3

from PIL import Image, ImageDraw
import os

def generate_icon(size):
    # Create image with gradient background
    img = Image.new('RGB', (size, size))
    draw = ImageDraw.Draw(img)

    # Draw gradient background (blue to purple)
    for y in range(size):
        ratio = y / size
        r = int(51 + ratio * (102 - 51))
        g = int(102 - ratio * (102 - 51))
        b = int(204 - ratio * (204 - 178))
        draw.line([(0, y), (size, y)], fill=(r, g, b))

    # Draw three overlapping photo frames
    frame_width = int(size * 0.5)
    frame_height = int(frame_width * 0.75)
    frame_thickness = max(2, int(size * 0.02))

    def draw_photo_frame(x, y, rotation=0):
        # Create a temporary image for the frame
        frame_img = Image.new('RGBA', (frame_width + 20, frame_height + 20), (0, 0, 0, 0))
        frame_draw = ImageDraw.Draw(frame_img)

        offset = 10
        # White frame border
        frame_draw.rectangle(
            [offset, offset, offset + frame_width, offset + frame_height],
            fill='white',
            outline='white'
        )

        # Dark image area
        frame_draw.rectangle(
            [offset + frame_thickness, offset + frame_thickness,
             offset + frame_width - frame_thickness, offset + frame_height - frame_thickness],
            fill=(76, 76, 76),
            outline=(76, 76, 76)
        )

        # Draw simple mountain shape
        img_x = offset + frame_thickness
        img_y = offset + frame_thickness
        img_w = frame_width - 2 * frame_thickness
        img_h = frame_height - 2 * frame_thickness

        mountain = [
            (img_x, img_y + img_h),
            (img_x + img_w * 0.3, img_y + img_h * 0.4),
            (img_x + img_w * 0.5, img_y + img_h * 0.7),
            (img_x + img_w * 0.7, img_y + img_h * 0.5),
            (img_x + img_w, img_y + img_h)
        ]
        frame_draw.polygon(mountain, fill=(128, 128, 128))

        # Draw sun/moon
        sun_radius = int(img_w * 0.12)
        sun_x = img_x + img_w * 0.75 - sun_radius
        sun_y = img_y + img_h * 0.25 - sun_radius
        frame_draw.ellipse(
            [sun_x, sun_y, sun_x + sun_radius * 2, sun_y + sun_radius * 2],
            fill=(178, 178, 178)
        )

        # Rotate if needed
        if rotation != 0:
            frame_img = frame_img.rotate(rotation, expand=False)

        # Paste onto main image
        img.paste(frame_img, (x - 10, y - 10), frame_img)

    # Draw frames at different positions with slight rotations
    center_x = size // 2
    center_y = size // 2

    draw_photo_frame(int(center_x - frame_width * 0.2), int(center_y - frame_height * 0.15), -8)
    draw_photo_frame(int(center_x - frame_width * 0.1), int(center_y - frame_height * 0.1), 6)
    draw_photo_frame(center_x - frame_width // 2, center_y - frame_height // 2, 0)

    return img

# Generate all required sizes
sizes = [
    ("icon_16x16.png", 16),
    ("icon_16x16@2x.png", 32),
    ("icon_32x32.png", 32),
    ("icon_32x32@2x.png", 64),
    ("icon_128x128.png", 128),
    ("icon_128x128@2x.png", 256),
    ("icon_256x256.png", 256),
    ("icon_256x256@2x.png", 512),
    ("icon_512x512.png", 512),
    ("icon_512x512@2x.png", 1024)
]

output_path = "slidey/Assets.xcassets/AppIcon.appiconset"

for filename, size in sizes:
    icon = generate_icon(size)
    filepath = os.path.join(output_path, filename)
    icon.save(filepath, 'PNG')
    print(f"Generated: {filename}")

print("Icon generation complete!")

File: tools/bin/generate_icon.swift
#!/usr/bin/swift

import AppKit
import CoreGraphics

func generateIcon(size: CGSize) -> NSImage {
    let image = NSImage(size: size)
    image.lockFocus()

    guard let context = NSGraphicsContext.current?.cgContext else {
        image.unlockFocus()
        return image
    }

    let rect = CGRect(origin: .zero, size: size)
    let cornerRadius = size.width * 0.2

    // Background gradient (blue to purple)
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    let colors = [
        NSColor(red: 0.2, green: 0.4, blue: 0.8, alpha: 1.0).cgColor,
        NSColor(red: 0.4, green: 0.2, blue: 0.7, alpha: 1.0).cgColor
    ] as CFArray

    if let gradient = CGGradient(colorsSpace: colorSpace, colors: colors, locations: [0.0, 1.0]) {
        let path = NSBezierPath(roundedRect: rect, xRadius: cornerRadius, yRadius: cornerRadius)
        path.addClip()
        context.drawLinearGradient(gradient, start: CGPoint(x: 0, y: size.height), end: CGPoint(x: size.width, y: 0), options: [])
    }

    // Draw three overlapping photo frames to represent a slideshow
    let frameWidth = size.width * 0.5
    let frameHeight = frameWidth * 0.75
    let frameThickness = size.width * 0.02

    // Back frame (rotated left)
    context.saveGState()
    context.translateBy(x: size.width * 0.3, y: size.height * 0.35)
    context.rotate(by: -0.15)
    drawPhotoFrame(context: context, width: frameWidth, height: frameHeight, thickness: frameThickness)
    context.restoreGState()

    // Middle frame (rotated right)
    context.saveGState()
    context.translateBy(x: size.width * 0.4, y: size.height * 0.4)
    context.rotate(by: 0.1)
    drawPhotoFrame(context: context, width: frameWidth, height: frameHeight, thickness: frameThickness)
    context.restoreGState()

    // Front frame (straight)
    context.saveGState()
    context.translateBy(x: size.width * 0.35, y: size.height * 0.45)
    drawPhotoFrame(context: context, width: frameWidth, height: frameHeight, thickness: frameThickness)
    context.restoreGState()

    image.unlockFocus()
    return image
}

func drawPhotoFrame(context: CGContext, width: CGFloat, height: CGFloat, thickness: CGFloat) {
    let frameRect = CGRect(x: -width/2, y: -height/2, width: width, height: height)
    let imageRect = frameRect.insetBy(dx: thickness, dy: thickness)

    // White frame
    context.setFillColor(NSColor.white.cgColor)
    context.fill(frameRect)

    // Dark image area
    context.setFillColor(NSColor(white: 0.3, alpha: 1.0).cgColor)
    context.fill(imageRect)

    // Add a simple mountain/photo silhouette
    let mountainPath = NSBezierPath()
    mountainPath.move(to: CGPoint(x: imageRect.minX, y: imageRect.minY))
    mountainPath.line(to: CGPoint(x: imageRect.minX + imageRect.width * 0.3, y: imageRect.minY + imageRect.height * 0.6))
    mountainPath.line(to: CGPoint(x: imageRect.minX + imageRect.width * 0.5, y: imageRect.minY + imageRect.height * 0.3))
    mountainPath.line(to: CGPoint(x: imageRect.minX + imageRect.width * 0.7, y: imageRect.minY + imageRect.height * 0.5))
    mountainPath.line(to: CGPoint(x: imageRect.maxX, y: imageRect.minY))
    mountainPath.close()

    context.setFillColor(NSColor(white: 0.5, alpha: 1.0).cgColor)
    mountainPath.fill()

    // Sun/moon circle
    let sunRadius = imageRect.width * 0.12
    let sunCenter = CGPoint(x: imageRect.maxX - imageRect.width * 0.25, y: imageRect.maxY - imageRect.height * 0.25)
    context.setFillColor(NSColor(white: 0.7, alpha: 1.0).cgColor)
    context.fillEllipse(in: CGRect(x: sunCenter.x - sunRadius, y: sunCenter.y - sunRadius, width: sunRadius * 2, height: sunRadius * 2))
}

extension NSBezierPath {
    func fill() {
        self.fill()
    }
}

// Generate all required sizes
let sizes: [(String, CGFloat)] = [
    ("icon_16x16.png", 16),
    ("icon_16x16@2x.png", 32),
    ("icon_32x32.png", 32),
    ("icon_32x32@2x.png", 64),
    ("icon_128x128.png", 128),
    ("icon_128x128@2x.png", 256),
    ("icon_256x256.png", 256),
    ("icon_256x256@2x.png", 512),
    ("icon_512x512.png", 512),
    ("icon_512x512@2x.png", 1024)
]

let outputPath = "slidey/Assets.xcassets/AppIcon.appiconset"

for (filename, size) in sizes {
    let icon = generateIcon(size: CGSize(width: size, height: size))

    if let tiffData = icon.tiffRepresentation,
       let bitmap = NSBitmapImageRep(data: tiffData),
       let pngData = bitmap.representation(using: .png, properties: [:]) {
        let filePath = "\(outputPath)/\(filename)"
        try? pngData.write(to: URL(fileURLWithPath: filePath))
        print("Generated: \(filename)")
    }
}

print("Icon generation complete!")
